### PR TITLE
add example for R.24

### PR DESCRIPTION
**Pull request by rknuus**
taken from https://cpppatterns.com/patterns/weak-reference.html, where
it's published under CC0 Public Domain Dedication

### CODE REVIEW

File: CppCoreGuidelines.md [hunk 1]
```diff
@@ -9151,7 +9151,37 @@ be able to destroy a cyclic structure.
...
+    {
+    public:
+      void set_back_reference(const std::weak_ptr<foo>& f)
+      {
+        this->back_reference = f;
+      }
+      void do_something()
+      {
+        std::shared_ptr<foo> shared_back_reference = this->back_reference.lock();
+        if (shared_back_reference) {
```
Comment (jwakely) [inline CppCoreGuidelines.md]: How about combining those two lines:

    if (auto shared_back_reference = back_reference.lock())

File: CppCoreGuidelines.md [hunk 2]
```diff
@@ -9151,7 +9151,37 @@ be able to destroy a cyclic structure.
...
+    private:
+      std::shared_ptr<bar> forward_reference;
+    };
+
+    class bar
+    {
+    public:
+      void set_back_reference(const std::weak_ptr<foo>& f)
+      {
+        this->back_reference = f;
```
Comment (cubbimew) [inline CppCoreGuidelines.md]: "this->" is kind of redundant

File: CppCoreGuidelines.md (hunk 1 above)
Comment (cubbimew) [inline CppCoreGuidelines.md]: `if(auto shared_back_reference = back_reference.lock())`, per ES.6: Declare names in for-statement initializers and conditions to limit scope

**Comment by AndrewPardoe** [comment]
Thanks! Merged the example.

**Pull request merged**